app/src/main/java/com/babylon/wallet/android/presentation/settings/preferences/assetshiding/HiddenAssetsViewModel.kt

```diff
@@ -0,0 +1,178 @@
+                                resource.address == assetAddress.v1.resourceAddress
+                            } ?: return@mapNotNull null
+                            val item = resource.items.firstOrNull { item ->
+                                item.localId == assetAddress.v1.nonFungibleLocalId
+                            }
+
+                            State.Asset(
+                                address = assetAddress,
+                                icon = item?.imageUrl ?: resource.iconUrl,
+                                name = resource.name,
+                                description = item?.name,
+                                hasDescription = true
+                            )
+                        },
+                        poolUnits = resources.pools.map { pool ->
+                            State.Asset(
+                                address = AssetAddress.PoolUnit(pool.address),
+                                icon = pool.metadata.keyImageUrl(),
+                                name = pool.name.takeIf { it.isNotBlank() },
+                                description = pool.associatedDApp?.name?.takeIf { it.isNotBlank() },
+                                hasDescription = true
+                            )
+                        }
+                    )
+                }
+                .flowOn(coroutineDispatcher)
+                .collect {
+                    _state.emit(it)
+                }
+        }
+    }
+
+    private suspend fun buildResources(
+        addresses: List<AssetAddress>
+    ): Resources {
+        val resourceAddresses = addresses.mapNotNull { address ->
+            when (address) {
+                is AssetAddress.Fungible -> address.v1
+                is AssetAddress.NonFungible -> address.v1.resourceAddress
+                else -> null
+            }
+        }
+
+        val resources = stateRepository.getResources(
+            addresses = resourceAddresses.toSet(),
+            underAccountAddress = null,
+            withDetails = true,
+            withAllMetadata = false
+        ).getOrThrow()
+
+        val nfts = addresses.filterIsInstance<AssetAddress.NonFungible>()
+            .groupBy { it.v1.resourceAddress }
+            .mapValues { entry ->
+                stateRepository.getNFTDetails(
+                    resourceAddress = entry.key,
+                    localIds = entry.value.map { it.v1.nonFungibleLocalId }.toSet()
+                ).getOrThrow()
+            }
+
+        return Resources(
+            fungibles = resources.filterIsInstance<Resource.FungibleResource>(),
+            nonFungibles = resources.filterIsInstance<Resource.NonFungibleResource>().map { collection ->
+                collection.copy(items = nfts[collection.address].orEmpty())
+            },
+            pools = stateRepository.getPools(
+                poolAddresses = addresses.pools().toSet()
+            ).getOrThrow()
```

Comment:
we do getOrThrow here and `buildResources` is used above inside `map {}`. Is there a chance of crash here? Maybe we should do mapCatching?

Reply:
Good catch, thank you!